hedera-mirror-rest/__tests__/utils.test.js

```
@@ -287,3 +287,268 @@ describe('Utils randomString tests', () => {
     expect(val).toMatch(/^[0-9a-z]{8}$/);
   });
 });
+
+const parseQueryParamTest = (testPrefix, testSpecs, parseParam) => {
+  testSpecs.forEach((testSpec) => {
+    test(`Utils parseAccountIdQueryParam - ${testSpec.name}`, () => {
+      const val = parseParam(testSpec);
+      expect(val[0]).toEqual(testSpec.expectedClause);
+      expect(val[1]).toEqual(testSpec.expectedValues);
+    });
+  });
+};
+
+//Common test names for parse*QueryParam tests
+const singleParamTestName = "Single parameter";
+const noParamTestName = "No parameters";
+const multipleParamsTestName = "Multiple parameters different ops";
+const extraParamTestName = "Extra useless parameter";
+const multipleEqualsTestName = "Multiple =";
+const duplicateParamsTestName = "Duplicate parameters"
+
+describe('Utils parseAccountIdQueryParam tests', () => {
+  const testSpecs = [
+    {
+      name: singleParamTestName,
+      parsedQueryParams: {"account.id": "gte:0.0.3"},
+      expectedClause: "account.id >= ?",
+      expectedValues: ["3"]
+    },
+    {
+      name: noParamTestName,
+      parsedQueryParams: {},
+      expectedClause: "",
+      expectedValues: []
+    },
+    {
+      name: multipleParamsTestName,
+      parsedQueryParams: {"account.id": ["gte:0.0.3", "lt:0.0.5", "2"]},
+      expectedClause: "account.id >= ? and account.id < ? and account.id IN (?)",
+      expectedValues: ["3", "5", "2"]
+    },
+    {
+      name: extraParamTestName,
+      parsedQueryParams: {
+        "account.id": "0.0.3",
+        "timestamp": "2000"
+      },
+      expectedClause: "account.id IN (?)",
+      expectedValues: ["3"]
+    },
+    {
+      name: multipleEqualsTestName,
+      parsedQueryParams: {"account.id": ["0.0.3", "4"]},
+      expectedClause: "account.id IN (?, ?)",
+      expectedValues: ["3", "4"]
+    },
+    {
+      name: duplicateParamsTestName,
+      parsedQueryParams: {"account.id": ["0.0.5", "5", "eq:0.0.5", "lte:0.0.3", "lte:0.0.3", "gte:0.0.3", "gte:0.0.4"]},
```

We should also probably throw an error if there's an eq/IN and a range operator since the eq/IN is more strict and would take precedence anyway in the query. Accepting both would only confuse people that this is not an OR. Throwing an error during client development time is better than production bugs caused by mismatched expectations. Probably applies to all query parameters.

---

I can see that, would just need to keep a map of ops seen during the parsing to check that.  The other potential valid confusion would be`account.id=lt:5&account.id=gt:10`, which would be a check for things outside of a given range, but would return no values because of the same issue (an AND, not an OR).

---

Did `account.id=lt:5&account.id=gt:10` use to be an OR and now it's an AND? If so, that's a breaking change for clients. We should consider making some operators an OR and some operators an AND and probably disallow mixing the two types.

---

No, all of the params that pass through `parseParams` before were parsed out individually and then joined together at the end with `return [partialQueries.join(' and '), values];`.  That was the cause of the original issue, where multiple = were ANDed together as literally "equals this AND that". This would just be another edge case that was never handled.

---

Will create a ticket to address this, it's an existing issue and probably deserves some attention on its own.

---

Created ticket #1654 to do validation on the query parameters as a group so that we can deal with these scenarios and few others I identified.